File: core/src/main/java/org/allenai/taggers/tag/CaseInsensitiveKeywordTagger.java
package org.allenai.taggers.tag;

import com.google.common.collect.Lists;
import org.allenai.nlpstack.core.ChunkedToken;
import org.allenai.nlpstack.core.Lemmatized;
import org.allenai.nlpstack.core.typer.Type;
import org.allenai.taggers.SentenceFunctions;
import org.allenai.taggers.StringFunctions;

import java.util.List;

/***
 * Search for exact keyword matches, ignoring case.
 * @author schmmd
 *
 */
public class CaseInsensitiveKeywordTagger extends KeywordTagger {
    public CaseInsensitiveKeywordTagger(String name, List<String> keywords) {
        super(name, Lists
                .transform(keywords, StringFunctions.toLowerCase));
    }

    /**
     * Constructor used by reflection.
     * @param name name of the tagger
     * @param args arguments to the tagger
     */
    public CaseInsensitiveKeywordTagger(String name, scala.collection.Seq<String> args) {
        this(name, scala.collection.JavaConversions.asJavaList(args));
    }

    @Override
    public List<Type> findTagsJava(final List<Lemmatized<ChunkedToken>> sentence) {
        return super.findTagsJava(SentenceFunctions.lowercase(sentence));
    }
}

File: core/src/main/scala/org/allenai/taggers/tag/CaseInsensitiveKeywordTagger.scala
package org.allenai.taggers.tag

import com.google.common.base.Function
import com.google.common.collect.Lists
import org.allenai.nlpstack.core.ChunkedToken
import org.allenai.nlpstack.core.Lemmatized
import org.allenai.nlpstack.core.typer.Type
import org.allenai.nlpstack.lemmatize.MorphaStemmer

/** Search for normalized keywords against a normalized sentence and tag the
  * match.  Lemmatized is defined by {@see Stemmer.lemmatize()}.
  * @author schmmd
  */
class CaseInsensitiveKeywordTagger(name: String, keywords: Seq[String])
extends KeywordTagger(name, keywords map (_.toLowerCase)) {
  override def canEqual(that: Any) = that.isInstanceOf[CaseInsensitiveKeywordTagger]

  override def tag(sentence: TheSentence, types: Seq[Type]): Seq[Type] = {
    keywordTag(sentence.strings map (_.toLowerCase), types)
  }
}
